[🔀 Member] 프로젝트 설정 및 소셜 회원가입/로그인 기능, Refresh 토큰 구현 -->

SeongGwangJu opened the pull request:
### 📅 2024.11.03 ~ 2024.11.05

### 🌵 Branch
feature/signin/#5 → develop

### 📢 Description
- 프로젝트 기본 설정 및 Security, JWT Config
- 응답 및 예외 통일
- 헬스체크 API
- Swagger 설정
- 회원가입 및 로그인 API
- 리프레쉬토큰으로 액세스토큰 재발급 API

### 💬 Issue Number
close: #5 

### ✔️ PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 코드가 정상적으로 동작하는지 테스트 했습니다.

### 🔖 Note
- 일단 회원가입시 닉네임은 UUID와 유사하게 랜덤으로 생성됩니다. (추후 수정 필요)

- 회원가입 겸 로그인
![image](https://github.com/user-attachments/assets/3852c79e-596a-4f40-8c3d-ff2cecfb5407)

- 리프레쉬토큰으로 액세스토큰 발급
![image](https://github.com/user-attachments/assets/4d9df02f-0d89-4082-8151-f1b081487269)

SeongGwangJu committed:
✨ feat: 헬스체크 API

SeongGwangJu committed:
✨ feat: 응답 및 예외 통일

- 응답/예외 통일
- 인증 예외 entry point 포함

SeongGwangJu committed:
✨ feat: Security 및 JWT Config

SeongGwangJu committed:
👷 chore: swagger 설정

SeongGwangJu committed:
✨ feat: 회원가입 및 로그인 기능 구현 - controller, service에는 액세스토큰 재발급 기능 포함

SeongGwangJu committed:
✨ feat: 액세스토큰 재발급 기능 구현

비즈니스 로직 및 컨트롤러는 이전 커밋에 포함

SeongGwangJu committed:
👷 chore: 프로젝트 기본 설정

SeongGwangJu committed:
:sparkles: feat: 사용자 인증 설정 파일 추가

for Security Configuration

iqveou6 reviewed (approved):
👍🏻👍🏻👍🏻👍🏻👍🏻👍🏻

gyudol commented on src/main/java/com/mulmeong/member/auth/application/AuthServiceImpl.java:
54행의 return respondSignIn(memberRepository.findByOauthId(requestDto.getOauthId()).get());에서
findByOauthId가 Optional 객체를 반환하는데 여기서 바로 get() 하면 Optional을 사용하는 의미가 사라지는 것 같습니다

.orElseThrow() 등으로 바꾸시는 것 제안드립니다!

SeongGwangJu replied:
respondSignIn(memberRepository.findByOauthId(requestDto.getOauthId()).get());
해당 코드는 이미 같은 조건으로 exist를 검증한 if문 안에서 사용하기때문에, 코드 간결함을 위해 orElseThrow()가 아닌 get을 사용하였습니다!

SeongGwangJu replied:
> 54행의 return respondSignIn(memberRepository.findByOauthId(requestDto.getOauthId()).get());에서 findByOauthId가 Optional 객체를 반환하는데 여기서 바로 get() 하면 Optional을 사용하는 의미가 사라지는 것 같습니다
> 
> .orElseThrow() 등으로 바꾸시는 것 제안드립니다!

respondSignIn(memberRepository.findByOauthId(requestDto.getOauthId()).get());
해당 코드는 이미 같은 조건으로 exist를 검증한 if문 안에서 사용하기때문에, 코드 간결함을 위해 orElseThrow()가 아닌 get을 사용하였습니다!

gyudol replied:
```
if (memberRepository.existsMemberByOauthIdAndOauthProvider(
                requestDto.getOauthId(), requestDto.getOauthProvider())) {

            return respondSignIn(memberRepository.findByOauthIdAndOauthProvider(
                    requestDto.getOauthId(), requestDto.getOauthProvider()).get());
```

```
return respondSignIn(memberRepository.findByOauthIdAndOauthProvider(
                    requestDto.getOauthId(), requestDto.getOauthProvider()).orElseThrow({예외_던질_내용}));
```

위의 경우는 DB를 2번 찌르게 되는데 다른 의도하신 게 있는지 궁금합니다.
아래 예시처럼 하면 exists로 검사할 필요가 없어져서 DB를 한 번만 찌르면 된다고 생각합니다!

SeongGwangJu replied:
👍 감사합니다 수정했습니다 
[⚡️ improve: 불필요한 DB 조회 및 반환 타입 개선](https://github.com/mul-meong/backend-member/pull/8/commits/789e50da1a404922350419532f700cd82a248bb2)

gyudol commented on src/main/java/com/mulmeong/member/auth/application/AuthServiceImpl.java:
// 지원하는 oAuth 플랫폼인지 확인
final Set<String> SupportedProviders = Set.of("kakao", "naver");

97행에서 메서드 안의 지역 변수로 정의해두면 메서드 들릴 때마다 Set을 새로 생성하므로 위에 전역변수로 정의하는 게 좋을 것 같습니다!
validOauth 메서드가 회원가입이나 로그인 할 때 마다 들리는 메서드이므로 많은 메서드 요청이 있을 거라 생각됩니다

SeongGwangJu replied:
👍 해당 부분 유지보수성 고려해 문자열이 아닌 enum으로 수정하였습니다.
https://github.com/mul-meong/backend-member/pull/8/files#diff-b7446947b39f41b69a4aac76deabaf23c38d08fcb2de8895ce9be14236b7ea35

gyudol commented on src/main/java/com/mulmeong/member/auth/application/AuthServiceImpl.java:
92행의 로직은 테스트를 위한 부분인 것 같아서 빼는 게 나을 것 같습니다!

// Swagger 테스트시 String으로 넘어오므로
if (oauthId.equals("string")) {
    return false;
}

SeongGwangJu replied:
> 92행의 로직은 테스트를 위한 부분인 것 같아서 빼는 게 나을 것 같습니다!
> 
> // Swagger 테스트시 String으로 넘어오므로 if (oauthId.equals("string")) { return false; }

반영하였습니다.
[♻️ refactor: oauth 검증 변경 방식 및 코드 간결화](https://github.com/mul-meong/backend-member/pull/8/commits/5cba090d47050b6fb44224cd781cf01267ab3bec)

SeongGwangJu committed:
♻️ refactor: oauth 검증 변경 방식 및 코드 간결화

- 불필요한 oauthId 검증 삭제
- oauthProvider 검증을 Enum으로 전환
- Oauth Id와 Provider 모두 일치하는 회원 조회로 변경(복합키로 변경)

gyudol commented on src/main/java/com/mulmeong/member/common/healthcheck/HealthCheckController.java:
```
+ ", port(local.server.port)=" + env.getProperty("local.server.port")
                    + ", port(server.port)=" + env.getProperty("server.port")
                    + ", gateway ip(env)=" + env.getProperty("gateway.ip"));
```

29행 ~ 31행 지우시는 것을 제안드립니다!
배포 서버에 올라간다고 생각했을 때 해당 내용들은 민감 정보들이므로 노출을 안 시키는 게 좋을듯 합니다.
health-check 메서드에서는 일반적으로 httpStatus.OK 정도만 던졌던 걸로 기억합니다

SeongGwangJu replied:
👍 감사합니다 수정했습니다 
[⚡️ improve: 불필요한 DB 조회 및 반환 타입 개선](https://github.com/mul-meong/backend-member/pull/8/commits/789e50da1a404922350419532f700cd82a248bb2)

Cualone commented on src/main/java/com/mulmeong/member/auth/infrastructure/MemberRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+    Optional<Member> findByOauthIdAndOauthProvider(String oauthId, String oauthProvider);
+
+    Boolean existsMemberByOauthIdAndOauthProvider(String oauthId, String oauthProvider);
```
boolean으로 쓰는 게 어떨까요

SeongGwangJu replied:
👍 감사합니다 수정했습니다 
[⚡️ improve: 불필요한 DB 조회 및 반환 타입 개선](https://github.com/mul-meong/backend-member/pull/8/commits/789e50da1a404922350419532f700cd82a248bb2)

Cualone reviewed (approved):
👍

SeongGwangJu committed:
⚡️ improve: 불필요한 DB 조회 및 반환 타입 개선

gyudol reviewed (approved):
고생하셨습니다 ✨
